dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriterImpl.kt

```diff
@@ -0,0 +1,260 @@
+import org.apache.arrow.vector.TinyIntVector
+import org.apache.arrow.vector.VarCharVector
+import org.apache.arrow.vector.VariableWidthVector
+import org.apache.arrow.vector.VectorSchemaRoot
+import org.apache.arrow.vector.types.DateUnit
+import org.apache.arrow.vector.types.FloatingPointPrecision
+import org.apache.arrow.vector.types.pojo.ArrowType
+import org.apache.arrow.vector.types.pojo.Field
+import org.apache.arrow.vector.types.pojo.FieldType
+import org.apache.arrow.vector.types.pojo.Schema
+import org.apache.arrow.vector.util.Text
+import org.jetbrains.kotlinx.dataframe.AnyCol
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.name
+import org.jetbrains.kotlinx.dataframe.api.convertToBigDecimal
+import org.jetbrains.kotlinx.dataframe.api.convertToBoolean
+import org.jetbrains.kotlinx.dataframe.api.convertToByte
+import org.jetbrains.kotlinx.dataframe.api.convertToDouble
+import org.jetbrains.kotlinx.dataframe.api.convertToFloat
+import org.jetbrains.kotlinx.dataframe.api.convertToInt
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalDate
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalDateTime
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalTime
+import org.jetbrains.kotlinx.dataframe.api.convertToLong
+import org.jetbrains.kotlinx.dataframe.api.convertToShort
+import org.jetbrains.kotlinx.dataframe.api.convertToString
+import org.jetbrains.kotlinx.dataframe.api.forEachIndexed
+import org.jetbrains.kotlinx.dataframe.api.map
+import org.jetbrains.kotlinx.dataframe.exceptions.CellConversionException
+import org.jetbrains.kotlinx.dataframe.exceptions.TypeConverterNotFoundException
+
+/**
+ * Save [dataFrame] content in Apache Arrow format (can be written to File, ByteArray, OutputStream or raw Channel) with [targetSchema].
+ * If [dataFrame] content does not match with [targetSchema], behaviour is specified by [mode], mismatches would be sent to [mismatchSubscriber]
+ */
+internal class ArrowWriterImpl(
+    override val dataFrame: DataFrame<*>,
+    override val targetSchema: Schema,
+    override val mode: ArrowWriter.Companion.Mode,
+    override val mismatchSubscriber: (ConvertingMismatch) -> Unit = ignoreMismatchMessage
+) : ArrowWriter {
+
+    private val allocator = RootAllocator()
+
+    private fun allocateVector(vector: FieldVector, size: Int) {
+        when (vector) {
+            is FixedWidthVector -> vector.allocateNew(size)
+            is VariableWidthVector -> vector.allocateNew(size)
+            else -> TODO("Not implemented for ${vector.javaClass.canonicalName}")
+        }
+    }
+
+    private fun infillWithNulls(vector: FieldVector, size: Int) {
+        when (vector) {
+            is BaseFixedWidthVector -> for (i in 0 until size) { vector.setNull(i) }
+            is BaseVariableWidthVector -> for (i in 0 until size) { vector.setNull(i) }
+            else -> TODO("Not implemented for ${vector.javaClass.canonicalName}")
```

Comment:
If you want to throw NotImplementedError, please do so explicitly, not via a TODO. Plus maybe an exception like IllegalArgumentException might be better here